File: py/lapcounter/new_lap.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from lapcounter import new_lap_pb2
from google.protobuf.json_format import MessageToJson, Parse


@dataclass
class NewLap:
    timestamp: int = 0
    
    _proto_message: new_lap_pb2.NewLap = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = new_lap_pb2.NewLap()

    def _populate_proto(self):
        self._proto_message.timestamp = self.timestamp

    @classmethod
    def _from_proto(cls, proto_message) -> "NewLap":
        return cls(
            timestamp = proto_message.timestamp,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "NewLap":
        message = new_lap_pb2.NewLap()
        message.ParseFromString(data)
        return cls(
            timestamp = message.timestamp,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "NewLap":
        message = new_lap_pb2.NewLap()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

File: src/lapcounter/new_lap.cpp
#include "lapcounter/new_lap.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace LapCounter
{
NewLap::NewLap(const PbLapCounter::NewLap& protobuf) {
    timestamp = protobuf.timestamp();
}

NewLap::operator PbLapCounter::NewLap() const {
    PbLapCounter::NewLap ret;
    ret.set_timestamp(timestamp);
    return ret;
}

std::string NewLap::serializeAsJsonString() const {
    PbLapCounter::NewLap protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string NewLap::serializeAsProtobufString() const {
    PbLapCounter::NewLap protobuf(*this);
    return protobuf.SerializeAsString();
}

bool NewLap::deserializeFromJsonString(const std::string& str) {
    PbLapCounter::NewLap protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool NewLap::deserializeFromProtobufString(const std::string& str) {
    PbLapCounter::NewLap protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}